src/components/ReportsPage/HeatmapPage/HeatmapUtils.ts

```diff
@@ -118,6 +129,29 @@ export const getAverageRuleScores = (
     .map(score => score / serviceCount);
 };
 
+export const getServicesInLevelsFromScores = (
+  ladderLevels: string[],
+  scores: ScorecardServiceScore[],
+): ScorecardServiceScore[][] => {
+  const groupedByLevels = groupReportDataBy(scores, 'ladderLevels');
+  return ladderLevels.map(level => groupedByLevels[level] ?? []);
+};
+
+export const getAverageRuleScoresByLevels = (
+  sortedRulesByLevels: string[],
+  scores: ScorecardServiceScore[],
+  serviceCount: number,
+): number[] => {
+  return scores
+    .map(score =>
+      getSortedRulesByLevelsFromScores(sortedRulesByLevels, score).map(
+        rule => rule.score,
+      ),
+    )
+    .reduce((r, a) => a.map((b, i) => (r[i] || 0) + (b ? 1 : 0)), [])
```

Comment:
ugh, this is hard to read, can we at least extract the mapper function and give it some meaningful name?